### PR TITLE
Fix NullReferenceException when accessing not existing section

### PR DESCRIPTION
**Pull request by caverna**
https://github.com/rickyah/ini-parser/issues/127

### CODE REVIEW

**Comment by rickyah** [comment]
Thanks for the PR. I might fix some style issues but looks good in a quick review. I'll try to integrate it this week now that I'm back from vacation ;)

**Comment by caverna** [comment]
Do you you want some help?
Something like a cleanup of tabs vs spaces?
I do suggest to use space instead tabs, because it's default behaviour for VSC# and Resharper... What do you thing?
I can do that without any compromisse from your side to include in master tree...
Later, we could refactor the code, removing a lot of region...
INI-Parser saved me a lot of time, many times. I think it's time to _pay back_ something... :)

**Comment by rickyah** [comment]
Oh, no, there's no way I'm going to fall into a tab vs spaces discussion 😆 I don't really mind, I use spaces in this project, if you find any tab is a mistake.

But when I talked about style issues I was referring to some whitespaces, I'll point it out in the PR :)

Glad to hear that this small library helped you! 👍

File: src/IniFileParser/Model/KeyDataCollection.cs [hunk 1]
```diff
@@ -83,14 +83,14 @@ public KeyDataCollection(KeyDataCollection ori, IEqualityComparer<string> search
                 if (_keyData.ContainsKey(keyName))
                     return _keyData[keyName].Value;
 
-                return null;
+                return string.Empty;
```
Comment (rickyah) [inline src/IniFileParser/Model/KeyDataCollection.cs]: Same as above with the null returns

File: src/IniFileParser/Model/IniData.cs [hunk 2]
```diff
@@ -41,9 +40,9 @@ public IniData(SectionDataCollection sdc)
             SectionKeySeparator = '.';
         }
 
-        public IniData(IniData ori): this((SectionDataCollection) ori.Sections)
+        public IniData(IniData ori) : this((SectionDataCollection)ori.Sections)
```
Comment (rickyah) [inline src/IniFileParser/Model/IniData.cs]: Please remove whitespace before the colon

File: src/IniFileParser.Tests/Unit/Model/SectionDataTests.cs [hunk 3]
```diff
@@ -116,7 +115,7 @@ public void try_removing_non_existing_key_from_section()
             Assert.That(sd.Keys.ContainsKey("asdf"), Is.False);
         }
 
-        [Test]
+        [Test, Ignore("The default behaviour of INI should be empty string, not null")]
```
Comment (rickyah) [inline src/IniFileParser.Tests/Unit/Model/SectionDataTests.cs]: See below, but you should unignore this as we need to revert to returning a null value

Also, next time, instead of ignoring the test will be nice to change it to test the new behaviour 😄

File: src/IniFileParser/Model/IniData.cs [hunk 4]
```diff
@@ -90,9 +89,9 @@ public IniParserConfiguration Configuration
         {
             get
             {
-                if (_sections.ContainsSection(sectionName)) return _sections[sectionName];
+                if (!_sections.ContainsSection(sectionName)) _sections.AddSection(sectionName);
 
-                return null;
```
Comment (rickyah) [inline src/IniFileParser/Model/IniData.cs]: Sorry, you need to put back the null return value. Returning an empty string is fine, and probably safer. But changing the return value qualifies as a breaking change, and I don't want to introduce such a change in a minor fix.

File: src/IniFileParser/Model/IniData.cs [hunk 5]
```diff
@@ -157,7 +156,7 @@ public void ClearAllComments()
         {
             Global.ClearComments();
 
-            foreach(var section in Sections)
+            foreach (var section in Sections)
```
Comment (rickyah) [inline src/IniFileParser/Model/IniData.cs]: please remove the space after the foreach

File: src/IniFileParser/Model/IniData.cs [hunk 6]
```diff
@@ -278,10 +277,10 @@ private void MergeSection(SectionData otherSection)
         /// </summary>
         private void MergeGlobal(KeyDataCollection globals)
         {
-            foreach(var globalValue in globals)
+            foreach (var globalValue in globals)
```
Comment (rickyah) [inline src/IniFileParser/Model/IniData.cs]: please remove the space after the foreach

File: src/IniFileParser/Model/KeyDataCollection.cs [hunk 7]
```diff
@@ -83,14 +83,14 @@ public KeyDataCollection(KeyDataCollection ori, IEqualityComparer<string> search
                 if (_keyData.ContainsKey(keyName))
                     return _keyData[keyName].Value;
 
-                return null;
```
Comment (rickyah) [inline src/IniFileParser/Model/KeyDataCollection.cs]: go back to returning null

File: src/IniFileParser/Model/IniData.cs [hunk 8]
```diff
@@ -121,7 +120,7 @@ public override string ToString()
         {
             return ToString(new DefaultIniDataFormatter(Configuration));
         }
-
```
Comment (rickyah) [inline src/IniFileParser/Model/IniData.cs]: nice catch! ;)

File: src/IniFileParser.Tests/Unit/Model/SectionDataTests.cs [hunk 9]
```diff
@@ -1,7 +1,6 @@
-﻿using System;
-
-using IniParser.Model;
+﻿using IniParser.Model;
 using NUnit.Framework;
+using System;
```
Comment (rickyah) [inline src/IniFileParser.Tests/Unit/Model/SectionDataTests.cs]: no need to move the using clauses around

File: src/IniFileParser/Model/IniData.cs [hunk 10]
```diff
@@ -1,7 +1,6 @@
-using System;
-using System.Net;
 using IniParser.Model.Configuration;
 using IniParser.Model.Formatting;
+using System;
```
Comment (rickyah) [inline src/IniFileParser/Model/IniData.cs]: why was this moved? An automatic formatter maybe?

**Comment by caverna** [comment]
I didn't thought that a quick fix could generate a huge break... :-D
About the spaces, maybe Resharper introduced those, I can check that later.
Besides that, I think that there's no reason approve this PR.
Don't worry, I won't take it bad! :-)

**Comment by rickyah** [comment]
We can leave this PR opened, I'll commit some changes when I have time and keep your fix ;)

**Comment by caverna** [comment]
Ok then, if you insist... :-)

**Comment by caverna** [comment]
I was thinking, about this compatibility break...
What do you think to put another field in the settings to control this behaviour?

**Comment by rickyah** [comment]
I'm closing this one and merging #130 that contains all @caverna commits except for the last one

**Comment by caverna** [comment]
Thank you! 👍
